New Foxx docs content

pluma opened the pull request:
This adds a section "guides" to the Foxx docs to cover common use cases and explain certain aspects of Foxx in more detail. It is intended to complement the existing API reference which is more descriptive and formal in nature.

~~NOTE: This PR likely contains a large number of broken links because I had to move a few chapters around. If someone can tell me how I can find and fix these myself I'll amend the PR accordingly.~~

This PR omits some chapters I originally planned to write but had to postpone to speed things up:

* Multipart (i.e. handling multipart requests)
* TypeScript (partially covered by the webpack chapter)
* Error handling (i.e. how to create proper error responses for production)
* Troubleshooting (diagnosing problems and debugging in lieu of a real debugger)

I still plan to write these chapters but they might not make it into 3.4 and will be added in a separate PR.

Simran-B commented on Documentation/Books/Manual/SUMMARY.md:
```diff
@@ -130,36 +130,45 @@
   * [Replication](Architecture/Replication/README.md)
   * [Write-ahead log](Architecture/WriteAheadLog.md)
-* [Foxx Microservices](Foxx/README.md)
+* [Foxx services](Foxx/README.md)
```
@pluma Is this approved by management/marketing?

pluma replied:
Good point, I'll revert that. I initially changed this because most of the docs just use "services" but I guess for the chapter title we should stick with it unless there's a deliberate decision. Thanks for catching this.

Simran-B commented:
All links and anchors fixed according to our build script. It might have missed some because of https://github.com/arangodb/planning/issues/2696 however

pluma commented:
@Simran-B is your review complete then or is it subject to arangodb/planning#2696 being fixed first?

Simran-B commented:
@pluma As long as building doesn't fail it is fine, but we might have to fix more links once all of them are tested. That should happen in another PR however. I did not review the actual content yet, nor did I reflow the text to ~80 chars per line.

pluma commented:
@Simran-B Do we have a consistent 80 chars per line rule now? We didn't when I last wrote documentation so I wasn't aware of that. I'll adjust the formatting accordingly.

@mpv1989 Can you review this PR for content soon-ish?

mpv1989 commented:
@pluma I have read about 80%. If the quality does not decrease I have nothing to criticize. For me there is already a go to merge this PR

pluma commented:
I've wrapped the lines so they're below the 80 character mark where possible. I've also fixed the missing links. And I've stashed the WIP chapters that didn't make it into this PR.

@Simran-B since Mark has reviewed the content, can you give the go-ahead if everything is fine?

pluma commented:
@Simran-B can you mark this as `Ready to Merge` if your review is complete?

Simran-B commented:
@pluma 

1. The Guide overview page could need some kind of grouping of related topics with a level 2 headline per group, and the linked articles should be presented as list.
  - _Bundling node modules_ and _linking services together_ seem related
  - _Development mode_, _Testing your services_ and maybe some other could as a group about developing services?
  - Maybe a headline per development stage? Testing and deployment in cluster would be the last stage I guess. Development mode, using collections, writing queries are rather early topics. Auth, file handling etc. are kind of advanced stuff.
  - Should the chapter `Foxx/Deployment.md` be moved to Guides perhaps? @sleto-it 
2. I fixed a few links to the HTTP book. Note that links within a book should use `.md` extensions, whereas links to other books have to end in `.html` because it will be used verbatim and we must refer to the static files as generated by the build script.
3. There is no strict 80 chars per line rule, but new content should follow this guideline to avoid unnecessarily big diffs and conflicts. Links are exempt.
4. I will quickly review the content and when everything is ready do another test build. If it succeeds then I will mark the PR as ready to merge.

Simran-B commented:
This sounds like you should also store binary data in the DB:

> ArangoDB is primarily a database. In most cases the best place to store data
is therefore inside the database, not on the file system.

Simran-B commented:
In `Guides/Dependencies.md`, is the following correct?

```
exports.login = (username, password) => {
...
exports.setPassword = (user, password) => {
```

or should this be `module.exports`?

pluma commented on Documentation/Books/Manual/Foxx/Guides/Dependencies.md:
```diff
@@ -88,12 +88,12 @@ Like [configuration](../Reference/Configuration.md),
 the [web interface](../../Programs/WebInterface/Services.md),
 the [Foxx CLI](../../Programs/FoxxCLI/README.md) or
-the [Foxx HTTP API](../../../HTTP/Foxx/Configuration.md).
+the [Foxx HTTP API](../../../HTTP/Foxx/Configuration.html).
```
Why is this one `.html`?

pluma replied:
Nevermind, see my question/response.

pluma commented:
> In Guides/Browser.md, is this actually meant to be a string of a Boolean?

Yes. It's a header value and header values must be strings.

> In Guides/Dependencies.md, is the following correct?

Yes, `module.exports` is only used when the "default" export should be something other than an object (e.g. a function or an instance of something).

> This sounds like you should also store binary data in the DB (Guides/Files.md):

Yes, however last I heard from @fceller was that this is "okay". I'm +1 on adding a warning to the extent of "ArangoDB is not a file storage service, use something sane like S3 if you need to routinely store files. Also we don't recommend storing files larger than X bytes" but I'm the wrong person to decide on the limits.

Storing data on the filesystem is a bad idea due to resilience/consistency, storing binary data in ArangoDB is a bad idea due to performance/scaling. I'd rather see users complain that they're killing their DB performance by being dumb than losing data and seeing difficult to diagnose bugs. We should definitely emphasise that you probably don't want to use ArangoDB to store your files but considering this is a recurring user need, we need to address this or people will rely on the filesystem and make things worse.

> The Guide overview page could need some kind of grouping of related topics with a level 2 headline per group, and the linked articles should be presented as list.

I'm not really sure how the topics can be grouped as most of them are fairly distinct. Expanding the list to prose was my attempt to work around that. If you see a meaningful way to group the topics, feel free to propose it.

> Bundling node modules and linking services together seem related

They're really not at all, they just both use the word "dependencies", which is why they refer to each other to disambiguate the term. They used to be handled in a single page previously which led to some confusion.

> Should the chapter Foxx/Deployment.md be moved to Guides perhaps?

I guess? But the chapter is mostly a reference to other chapters, which are also referenced elsewhere as appropriate. I also don't see an easy way to expand the content.

> links to other books have to end in .html because it will be used verbatim and we must refer to the static files as generated by the build script.

That means everything on the same level as `Manual`, correct?

> What about the screenshot placeholders? Should we create and add them all, or do this later?

I don't think I added any new placeholders. The existing ones were originally intended for 3.0 and was initially postponed because the design wasn't finalised at the time. I'm not entirely sure they're needed in most places anymore but they might help with the getting started tutorial.

---

On an unrelated note: if you have questions regarding code that's part of the PR, please make the comment inline as it helps responding. There's no reply feature in GitHub so I have to quote messages individually via copy-paste in cases like this.

Simran-B commented on Documentation/Books/Manual/Foxx/Guides/DevelopmentMode.md:
```diff
@@ -0,0 +1,75 @@
+Additionally, error responses generated by Foxx will include stacktraces and
+when viewed in a browser may include relevant sections of the service code
+that generated the error.
+
```
We should explain somewhere here how to find the folder where the Foxx services are in. Is there a way to find out via arangosh? Otherwise, one would have to look at arangod.conf and possibly startup parameters specified on the command line that overwrite the default config.

pluma replied:
Fixed in https://github.com/arangodb/arangodb/pull/5842/commits/4d7b0b47c8e76d69a640bbe1691d928fb753317a

pluma commented:
I've rebased to fix the conflicts introduced in #5904.
I've also added a comment about where the service source code is located and expanded the file handling chapter with a few warnings:

https://github.com/arangodb/arangodb/pull/5842/commits/d75142aae35bb4b84f141b459e34937a8e04103f

pluma commented:
@Simran-B Can you review these two changes to see whether this addresses the concerns raised on slack?

pluma commented:
I've also just added two info boxes to call out the distinction between node and Foxx dependencies in the corresponding chapters, hopefully making it easier to find.

pluma commented:
@Simran-B if there's nothing left to do, please flag this as ready to merge.

pluma commented:
I've rebased to devel so hopefully the docs can be built.

pluma commented:
I've added some grouping and what I think comes close to a logical progression. The third group is mostly a catch-all for various footguns (i.e. "we don't really advertise these things but if you absolutely have to do them, here's how"). Feel free to suggest a better naming if you're unhappy with the headlines.

Jenkins should push updates to https://docs.arangodb.com/review/documentation/foxx-new-content/Manual/Foxx/Guides/

pluma commented:
:tada:
